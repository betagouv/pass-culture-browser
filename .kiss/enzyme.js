import React from 'react'
import { shallow } from 'enzyme'

import MyComponent from '../MyComponent'

describe('src | components | MyComponent', () => {
  describe('snapshot', () => {
    it('should match snapshot', () => {
      const props = { errors: {} }
      const wrapper = shallow(<MyComponent {...props} />)
      expect(wrapper).toBeDefined()
      expect(wrapper).toMatchSnapshot()
    })
  })
})
