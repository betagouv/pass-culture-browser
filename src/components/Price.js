import React from 'react'

const Price = ({ value, free, Tag, className }) => {
  return (
    <Tag className={className}>{
      value === 0
        ? free
        : value && (value.toString().replace('.', ',')+'â‚¬')
    }</Tag>
  )
}

Price.defaultProps = {
  free: 'gratuit',
  Tag: 'span',
  className: 'price',
}

export default Price;
