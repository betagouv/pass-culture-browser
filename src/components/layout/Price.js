import React from 'react'
import PropTypes from 'prop-types'

import { getDisplayPrice } from '../../helpers'

const Price = ({ free, value, className, ...rest }) => {
  const price = getDisplayPrice(value, free)
  const classes = `price ${className}`
  return (
    <span {...rest} className={classes}>
      {price}
    </span>
  )
}

Price.defaultProps = {
  className: '',
  free: null,
  value: null,
}

Price.propTypes = {
  className: PropTypes.string,
  free: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.array, PropTypes.number]),
}

export default Price
