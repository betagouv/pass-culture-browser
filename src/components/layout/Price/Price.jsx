import React from 'react'
import PropTypes from 'prop-types'

import { getDisplayPrice } from '../../../helpers'

const Price = ({ free, value, className, ...rest }) => {
  const price = getDisplayPrice(value, free)

  return (
    <span
      {...rest}
      className={className}
    >
      {price}
    </span>
  )
}

Price.defaultProps = {
  className: '',
  free: null,
  value: null,
}

Price.propTypes = {
  className: PropTypes.string,
  free: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.array, PropTypes.number]),
}

export default Price
