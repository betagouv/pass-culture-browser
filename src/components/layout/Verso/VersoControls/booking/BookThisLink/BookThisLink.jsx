import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'

import VersoPriceFormatter from '../VersoPriceFormatter/VersoPriceFormatter'
import Price from '../../../../Price'

class BookThisLink extends Component {
  formatOutputPrice = ([startingPrice, endingPrice], devise) => (
    <VersoPriceFormatter
      devise={devise}
      endingPrice={endingPrice}
      startingPrice={startingPrice}
    />
  )

  render() {
    const { isFinished, priceRange, destinationLink } = this.props
    return (
      <Link
        className="flex-columns is-bold is-white-text fs18"
        disabled={isFinished}
        id="verso-booking-button"
        to={destinationLink}
      >
        <Price
          className="pc-ticket-button-price flex-columns items-center"
          format={this.formatOutputPrice}
          free="Gratuit"
          value={priceRange}
        />
        <span className="pc-ticket-button-label">{'Jâ€™y vais !'}</span>
      </Link>
    )
  }
}

BookThisLink.defaultProps = {
  isFinished: false,
}

BookThisLink.propTypes = {
  destinationLink: PropTypes.string.isRequired,
  isFinished: PropTypes.bool,
  priceRange: PropTypes.arrayOf(PropTypes.number.isRequired).isRequired,
}

export default BookThisLink
