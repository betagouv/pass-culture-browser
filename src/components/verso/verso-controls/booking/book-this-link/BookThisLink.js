import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'

import Price from '../../../../layout/Price'
import VersoPriceFormatter from '../verso-price-formatter/VersoPriceFormatter'

class BookThisLink extends Component {
  formatOutputPrice = ([startingPrice, endingPrice], devise) => (
    <VersoPriceFormatter
      devise={devise}
      endingPrice={endingPrice}
      startingPrice={startingPrice}
    />
  )

  render() {
    const { isFinished, linkDestination, priceValue } = this.props

    return (
      <Link
        className="flex-columns is-bold is-white-text fs18"
        disabled={isFinished}
        id="verso-booking-button"
        to={linkDestination}
      >
        <Price
          className="pc-ticket-button-price flex-columns items-center"
          format={this.formatOutputPrice}
          free="Gratuit"
          value={priceValue}
        />
        <span className="pc-ticket-button-label">{'Jâ€™y vais !'}</span>
      </Link>
    )
  }
}

BookThisLink.defaultProps = {
  isFinished: false,
}

BookThisLink.propTypes = {
  isFinished: PropTypes.bool,
  linkDestination: PropTypes.string.isRequired,
  priceValue: PropTypes.arrayOf(PropTypes.number).isRequired,
}

export default BookThisLink
