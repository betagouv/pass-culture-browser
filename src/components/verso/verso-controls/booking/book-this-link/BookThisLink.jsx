import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'

import VersoPriceFormatter from '../verso-price-formatter/VersoPriceFormatter'
import Price from '../../../../layout/Price'
import { getPriceRangeFromStocks } from '../../../../../helpers'

class BookThisLink extends Component {
  formatOutputPrice = ([startingPrice, endingPrice], devise) => (
    <VersoPriceFormatter
      devise={devise}
      endingPrice={endingPrice}
      startingPrice={startingPrice}
    />
  )

  getLinkDestination = () => {
    const { match, location } = this.props
    const { search } = location
    const { params, url } = match
    if (params.bookings) {
      return url
    }
    const isValid = url && typeof url === 'string'
    if (!isValid) {
      throw new Error('Invalid url parameter')
    }
    let formattedUrl = url
    return `${formattedUrl}/reservations${search || ''}`
  }

  render() {
    const { isFinished, recommendation } = this.props
    const { offer } = recommendation
    const { stocks } = offer
    const price = getPriceRangeFromStocks(stocks)

    return (
      <Link
        className="flex-columns is-bold is-white-text fs18"
        disabled={isFinished}
        id="verso-booking-button"
        to={this.getLinkDestination()}
      >
        <Price
          className="pc-ticket-button-price flex-columns items-center"
          format={this.formatOutputPrice}
          free="Gratuit"
          value={price}
        />
        <span className="pc-ticket-button-label">{'Jâ€™y vais !'}</span>
      </Link>
    )
  }
}

BookThisLink.defaultProps = {
  isFinished: false,
}

BookThisLink.propTypes = {
  isFinished: PropTypes.bool,
  location: PropTypes.shape({
    search: PropTypes.string.isRequired,
  }).isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      bookings: PropTypes.string
    }).isRequired,
    url: PropTypes.string.isRequired,
  }).isRequired,
  recommendation: PropTypes.shape({
    offer: PropTypes.shape({
      stocks: PropTypes.arrayOf(PropTypes.shape({
        price: PropTypes.number
      })).isRequired
    }).isRequired
  }).isRequired
}

export default BookThisLink
