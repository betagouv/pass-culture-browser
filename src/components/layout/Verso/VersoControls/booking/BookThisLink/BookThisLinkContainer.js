import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import { compose } from 'redux'

import BookThisLink from './BookThisLink'
import getPriceRangeFromStocks from '../../../../../../helpers/getPriceRangeFromStocks'
import selectIsFinishedByMatch from '../../../../../../selectors/selectIsFinishedByMatch'
import selectOfferByMatch from '../../../../../../selectors/selectOfferByMatch'

export const mapStateToProps = (state, ownProps) => {
  const { match } = ownProps
  const isFinished = selectIsFinishedByMatch(state, match)
  const offer = selectOfferByMatch(state, match) || {}
  const { stocks } = offer
  const priceRange = getPriceRangeFromStocks(stocks)
  return {
    isFinished,
    priceRange
  }
}

export default compose(
  withRouter,
  connect(mapStateToProps)
)(BookThisLink)
