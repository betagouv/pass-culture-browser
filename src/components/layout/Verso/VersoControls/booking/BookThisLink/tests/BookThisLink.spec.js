import React from 'react'
import { mount, shallow } from 'enzyme'
import { MemoryRouter } from 'react-router-dom'

import Price from '../../../../../Price'
import BookThisLink from '../BookThisLink'

describe('src | components | verso | verso-controls | booking | BookThisLink', () => {
  it('should match snapshot with required props', () => {
    // given
    const props = {
      isFinished: false,
      location: {
        search: '',
      },
      match: {
        params: {},
        url: '/decouverte',
      },
      priceRange: [10, 30],
    }

    // when
    const wrapper = shallow(<BookThisLink {...props} />)

    // then
    const buttonLabel = wrapper.find('.pc-ticket-button-label')
    expect(wrapper).toBeDefined()
    expect(wrapper).toMatchSnapshot()
    expect(buttonLabel).toHaveLength(1)
    expect(buttonLabel.text()).toBe('J’y vais !')
  })

  it('should render Gratuit label when price value is 0', () => {
    // given
    const props = {
      isFinished: false,
      location: {
        search: '',
      },
      match: {
        params: {},
        url: '/decouverte',
      },
      priceRange: [0],
    }

    // when
    const wrapper = mount(
      <MemoryRouter>
        <BookThisLink {...props} />
      </MemoryRouter>
    )

    // then
    const priceComponent = wrapper.find(Price)
    expect(priceComponent).toHaveLength(1)
    expect(priceComponent.hasClass('pc-ticket-button-price')).toBe(true)
    const price = wrapper.find('.price')

    // then
    expect(price).toHaveLength(1)
    expect(price.text()).toStrictEqual('Gratuit')
  })

  it('should render a price range when multiples prices are given', () => {
    // given
    const nbsp = '\u00a0'
    const arrow = '\u27A4'
    const minPrice = 10
    const maxPrice = 30
    const props = {
      isFinished: false,
      location: {
        search: '',
      },
      match: {
        params: {},
        url: '/decouverte',
      },
      priceRange: [minPrice, maxPrice],
    }

    // when
    const wrapper = mount(
      <MemoryRouter>
        <BookThisLink {...props} />
      </MemoryRouter>
    )

    // then
    const priceComponent = wrapper.find(Price)
    expect(priceComponent).toHaveLength(1)
    expect(priceComponent.hasClass('pc-ticket-button-price')).toBe(true)

    // then
    const price = wrapper.find('.price')
    expect(price).toHaveLength(1)
    expect(price.text()).toStrictEqual(`${minPrice}${nbsp}${arrow}${nbsp}${maxPrice}${nbsp}€`)
  })

  describe('getLinkDestination', () => {
    it('should add reservations to current url without query to open booking card', () => {
      // given
      const mediationId = 'CG'
      const offerId = 'BF'
      const pathname = '/decouverte'
      const search = '?foo'
      const props = {
        isFinished: false,
        location: {
          pathname,
          search,
        },
        match: {
          params: {},
          url: `/decouverte/${offerId}/${mediationId}`,
        },
        priceRange: [],
      }
      const expected = `${pathname}/${offerId}/${mediationId}/reservation${search}`

      // when
      const wrapper = shallow(<BookThisLink {...props} />)
      const result = wrapper.instance().getLinkDestination()

      // then
      expect(result).toStrictEqual(expected)
    })

    it('should not add reservations to current url if bookings already in match params', () => {
      // given
      const mediationId = 'CG'
      const offerId = 'BF'
      const pathname = '/decouverte'
      const search = '?foo'
      const url = `http://${pathname}/${offerId}/${mediationId}/reservation${search}`
      const props = {
        isFinished: false,
        location: {
          pathname,
          search,
        },
        match: {
          params: {
            bookings: 'reservations',
          },
          url,
        },
        priceRange: [],
      }
      const expected = url

      // when
      const wrapper = shallow(<BookThisLink {...props} />)
      const result = wrapper.instance().getLinkDestination()

      // then
      expect(result).toStrictEqual(expected)
    })
  })
})
