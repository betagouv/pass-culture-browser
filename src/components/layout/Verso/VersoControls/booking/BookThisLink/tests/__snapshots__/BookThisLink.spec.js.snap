// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src | components | verso | verso-controls | booking | BookThisLink should match snapshot with required props 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BookThisLink
    isFinished={false}
    location={
      Object {
        "search": "",
      }
    }
    match={
      Object {
        "params": Object {},
        "url": "/decouverte",
      }
    }
    priceRange={
      Array [
        10,
        30,
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Price
          className="pc-ticket-button-price flex-columns items-center"
          format={[Function]}
          free="Gratuit"
          value={
            Array [
              10,
              30,
            ]
          }
        />,
        <span
          className="pc-ticket-button-label"
        >
          J’y vais !
        </span>,
      ],
      "className": "flex-columns is-bold is-white-text fs18",
      "disabled": false,
      "id": "verso-booking-button",
      "replace": false,
      "to": "/decouverte/reservations",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "className": "pc-ticket-button-price flex-columns items-center",
          "format": [Function],
          "free": "Gratuit",
          "value": Array [
            10,
            30,
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "J’y vais !",
          "className": "pc-ticket-button-label",
        },
        "ref": null,
        "rendered": "J’y vais !",
        "type": "span",
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Price
            className="pc-ticket-button-price flex-columns items-center"
            format={[Function]}
            free="Gratuit"
            value={
              Array [
                10,
                30,
              ]
            }
          />,
          <span
            className="pc-ticket-button-label"
          >
            J’y vais !
          </span>,
        ],
        "className": "flex-columns is-bold is-white-text fs18",
        "disabled": false,
        "id": "verso-booking-button",
        "replace": false,
        "to": "/decouverte/reservations",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "className": "pc-ticket-button-price flex-columns items-center",
            "format": [Function],
            "free": "Gratuit",
            "value": Array [
              10,
              30,
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "J’y vais !",
            "className": "pc-ticket-button-label",
          },
          "ref": null,
          "rendered": "J’y vais !",
          "type": "span",
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
